adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java

```diff
@@ -0,0 +1,141 @@
+                     List<Object> pStmtValList) {
+        super(jdbcSpace);
+
+        this.isReadStmt = isReadStmt;
+        this.pStmtSqlStr = pStmtSqlStr;
+        this.pStmtValList = pStmtValList;
+        this.isPreparedStmt = StringUtils.contains(pStmtSqlStr, "?");
+
+        // NOTE:
+        // - The write DML statement MUST use prepared statement!
+        // - The read DML statement can use either a prepared or a regular statement.
+        if (!isReadStmt && !isPreparedStmt) {
+            throw new JDBCAdapterInvalidParamException(
+                "Write DML statement must use prepared statement format!");
+        }
+
+        if (isPreparedStmt) {
+            int expectedFiledCnt = StringUtils.countMatches(pStmtSqlStr, "?");
+            int actualFieldCnt = pStmtValList.size();
+            if (expectedFiledCnt != actualFieldCnt) {
+                throw new JDBCAdapterUnexpectedException(
+                    "Provided value count (" + actualFieldCnt
+                        + ") doesn't match the expected field count (" + expectedFiledCnt
+                        + ") for the prepared statement: \"" + pStmtSqlStr + "\""
+                );
+            }
+        }
+    }
+
+    // Only applicable to a prepared statement
+    protected PreparedStatement setPrepStmtValues(PreparedStatement stmt, List<Object> valList) {
+        assert (stmt != null);
+
+        for (int i=0; i<valList.size(); i++) {
+            int fieldIdx = i + 1;
+            String inputFieldVal = ((String)valList.get(i)).trim();
+
+            try {
+                Object fieldValObj = inputFieldVal;
+
+                // If the 'fieldVal' is a string like "[<float_num_1>, <float_num_2>, ... <float_num_n>]" format,
+                // convert it to the Vector object
+                if ( inputFieldVal.startsWith("[") && inputFieldVal.endsWith("]") ) {
+                    JDBCPgVector vector = new JDBCPgVector();
+                    vector.setValue(inputFieldVal);
+                    fieldValObj = vector;
+                }
+
+                stmt.setObject(fieldIdx, fieldValObj);
+            }
+            catch (JDBCPgVectorException | SQLException e) {
+                throw new RuntimeException(
+                    "Failed to parse the prepared statement value for field[" + fieldIdx + "] " + inputFieldVal);
+            }
+        }
+
+        return stmt;
+    }
+
+    protected void processCommit() {
+        try {
+            if (!jdbcConnection.getAutoCommit()) {
+                jdbcConnection.commit();
+                LOGGER.debug(() -> LOG_COMMIT_SUCCESS);
+            }
+        } catch (SQLException e) {
+            throw new JDBCAdapterUnexpectedException("Failed to process JDBC statement commit!");
+        }
+    }
+
+    protected Statement createDMLStatement() {
+        Statement stmt = jdbcStmtTL.get();
+
+        try {
+            if (stmt == null) {
+                if (isPreparedStmt)
```

Comment:
Should we keep it like this instead?

```suggestion
                if (isPreparedStmt && StringUtils.isNotBlank(pStmtSqlStr))
```

Reply:
It shouldn't be an empty string. But for safety, add "assert(StringUtils.isNotBlank(pStmtSqlStr));" in the constructor.